gpu-simulator/main.cc

```diff
@@ -81,41 +81,41 @@ int main(int argc, const char **argv) {
-      // Read trace header info for window_size number of kernels
-      while (kernels_info.size() < window_size && i < commandlist.size()) {
-        kernel_trace_t* kernel_trace_info = tracer.parse_kernel_info(commandlist[i].command_string);
-        kernel_info = create_kernel_info(kernel_trace_info, m_gpgpu_context, &tconfig, &tracer);
-        kernels_info.push_back(kernel_info);
-        std::cout << "Header info loaded for kernel command : " << commandlist[i].command_string << std::endl;
+    //gulp up as many commands as possible - either cpu_gpu_mem_copy 
+    //or kernel_launch - until the vector "kernels_info" has exceeded
+    //the window_size or no command is left in commandlist
+    while (kernels_info.size() < window_size && i < commandlist.size()) {
+      trace_kernel_info_t *kernel_info = NULL;
+      if (commandlist[i].m_type == command_type::cpu_gpu_mem_copy) {
+        size_t addre, Bcount;
+        tracer.parse_memcpy_info(commandlist[i].command_string, addre, Bcount);
+        std::cout << "launching memcpy command : " << commandlist[i].command_string << std::endl;
+        m_gpgpu_sim->perf_memcpy_to_gpu(addre, Bcount);
         i++;
+      } else if (commandlist[i].m_type == command_type::kernel_launch) {
+        // Read trace header info for window_size number of kernels
+          kernel_trace_t* kernel_trace_info = tracer.parse_kernel_info(commandlist[i].command_string);
+          kernel_info = create_kernel_info(kernel_trace_info, m_gpgpu_context, &tconfig, &tracer);
+          kernels_info.push_back(kernel_info);
+          std::cout << "Header info loaded for kernel command : " << commandlist[i].command_string << std::endl;
+          i++;
       }
-      
-      // Launch all kernels within window that are on a stream that isn't already running
-      for (auto k : kernels_info) {
-        bool stream_busy = false;
-        for (auto s: busy_streams) {
-          if (s == k->get_cuda_stream_id())
-            stream_busy = true;
-        }
-        if (!stream_busy && m_gpgpu_sim->can_start_kernel() && !k->was_launched()) {
-          std::cout << "launching kernel name: " << k->get_name() << " uid: " << k->get_uid() << std::endl;
-          m_gpgpu_sim->launch(k);
-          k->set_launched();
-          busy_streams.push_back(k->get_cuda_stream_id());
-        }
+    }
+
+    // Launch all kernels within window that are on a stream that isn't already running
+    for (auto k : kernels_info) {
+      bool stream_busy = false;
+      for (auto s: busy_streams) {
+        if (s == k->get_cuda_stream_id())
+          stream_busy = true;
+      }
+      if (!stream_busy && m_gpgpu_sim->can_start_kernel() && !k->was_launched()) {
+        std::cout << "launching kernel name: " << k->get_name() << " uid: " << k->get_uid() << std::endl;
+        m_gpgpu_sim->launch(k);
+        k->set_launched();
+        busy_streams.push_back(k->get_cuda_stream_id());
       }
     }
-    else if (kernels_info.empty())
```

Comment:
I don't think you should remove the check for bad commands in the kernelslist.g file

Reply:
I agree. It's unwise to not check for bad commands, but the condition would need to be changed - no longer be "kernels_info.empty()", but something more like a "default" case i.e. if "commandlist[i].m_type" is neither kernel_launch nor cpu_gpu_mem_copy, then panic. 
Here's my understanding of why the condition for panicking used to be` if(kernels_info.empty())` - if we enter the outermost while loop with an empty `kernels_info`, and we reach the line `else if (kernels_info.empty())`, it implies `commandlist[i]` was a non-supported command type, so we should panic for undefined command. But that was not sufficient to catch all non-supported commands yet. What if we enter the outermost while loop with `kernels_info` not empty and fails to match the case for both `kernel_launch` and `cpu_gpu_mem_copy`, then the old code would not panic and quietly carry on executing.  
I added the code to check for unsupported command as a else{} statement, does that look better?

Reply:
looks good